nixos/systemd-confinement: support ProtectSystem=/DynamicUser=

ju1m opened the pull request:
## Description of changes
Use a `RuntimeDirectory=` as `RootDirectory=` instead of using `/var/empty` and `TemporaryFileSystem=/`.
See also [a previous discussion about this change](https://discourse.nixos.org/t/hardening-systemd-services/17147/14).

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [X] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Ping @aszlig as author of `systemd-confinement`.
Ping @martinetd as fervent user of `systemd-confinement`.
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ju1m committed:
nixos/systemd-confinement: support ProtectSystem=/DynamicUser=

See https://discourse.nixos.org/t/hardening-systemd-services/17147/14

martinetd commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -98,8 +98,9 @@ in {
         serviceConfig = {
-          RootDirectory = "/var/empty";
-          TemporaryFileSystem = "/";
+          RuntimeDirectory = [ "confinement/${mkPathSafeName name}" ];
```
I was wondering if any service use this for something already -- we might need to change this to modify the user's value e.g. confinement/name/run/oldvalue if there was one, so they'll have /run/oldvalue in the chroot
(note: after checking I have a couple of services with a RuntimeDirectory and confinement set (miniflux, postgresql), but nothing in there so they probably don't need it... I also have one that actually uses the directory (mpd), so I'll be able to report problem when testing in a bit)

martinetd replied:
(still not tested) this is actually probably fine, I tested multiple RuntimeDirectory with a path within the chroot dir but using the option normally works as a list, so as long as we're working with services that properly define this as a list it'll work properly.

I don't think there's much we can do if a service defines it as a string though (would need a more complex merge mechanism with mixed types somewhere and I don't know how to do that in nix), we'll need to fix such services when coming accross them...

martinetd replied:
Okay, that turned out quite painful as many of the services I use (postgres, mpd, miniflux, matrix-synapse...) define this as a string, and the merge policy fail when we have a list and a string: https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/systemd-unit-options.nix#L20

While I understand where that comes from and it's well intended I don't think it's actually that helpful - I guess if a user overrides some parameter with `serviceConfig.foo = bar` without override because the nixos service doesn't set it, and the nixos service starts adding one all of a sudden they'd be happy to get an error?...
So I'd be tempted to say we should just always concat+filter dups in there and wipe this under the rug, but that might not be satisfying for everyone.

I'm not sure, but as this I have just about as many overrides due to this than I had with dynamicUser = mkForce false, so we probably ought to come up with something here.

martinetd commented on nixos/tests/systemd-confinement.nix:
```diff
@@ -48,16 +48,22 @@ import ./make-test-python.nix {
+              # systemd’s setup_namespace() calls base_filesystem_create()
+              # which creates those usual top level directories.
+              # Check that they're empty.
+              paths = machine.succeed('chroot-exec ls -Am /dev /etc /proc /root /sys /usr /var | paste -sd,').strip()
```
(nitpick: if you need to change something at some point, find's output might be more stable e.g. just the list of directories)

ju1m replied:
I tried `find` but it was not available in the chroot and it was not immediately obvious to me how to put it in there, so I sticked with `ls`. Isn't `ls` output stable/sorted enough here?

martinetd replied:
I'd have expected it to be pulled in the root if you use it with `${pkgs.findutils}/bin/find`, but if you tried I guess this is good enough (I'm not really that worried ls output would change -- it's more of a "it's not really easy to parse with multiple directories content")

martinetd reviewed (commented):
Great, DynamicUser has been something I've had to override way too often to use this, I'll try to give it a try with my dozen services tomorrow!

martinetd commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -98,8 +98,9 @@ in {
-          RootDirectory = "/var/empty";
-          TemporaryFileSystem = "/";
+          RuntimeDirectory = [ "confinement/${mkPathSafeName name}" ];
+          RootDirectory = lib.mkForce "/run/confinement/${mkPathSafeName name}";
```
(discussed out-of-band)
I was under the impression that 'mkForce' is best left for users, so we'd probably want to use mkOverride 90 or some value between default value and mkForce for this, even if nobody should ever override this...

But it doesn't look like there's a well defined value for this, and I see plenty enough of mkForce in nixos/modules so this is probably fine? If a user *realy* wants to override this for some reason they can use an even lower priority.

For reference:
```
  mkOptionDefault = mkOverride 1500; # priority of option defaults
  mkDefault = mkOverride 1000; # used in config sections of non-user modules to set a default
  defaultOverridePriority = 100;
  mkImageMediaOverride = mkOverride 60; # image media profiles can be derived by inclusion into host config, hence needing to override host config, but do allow user to mkForce
  mkForce = mkOverride 50;
  mkVMOverride = mkOverride 10; # used by ‘nixos-rebuild build-vm’
```

aszlig replied:
Why does this even need to use `mkForce` in the first place?

martinetd reviewed (commented):
(as for testing, I've coupled this with a 1 month nixos-unstable update and both my systems fail to build in multiple places, so that'll take a little bit longer...)

aszlig commented on nixos/tests/systemd-confinement.nix:
```diff
@@ -48,16 +48,22 @@ import ./make-test-python.nix {
-              paths = machine.succeed('chroot-exec ls -1 / | paste -sd,').strip()
-              assert_eq(paths, "bin,nix,run")
+              paths = machine.succeed('chroot-exec ls -Am /').strip()
+              assert_eq(paths, "bin, dev, etc, nix, proc, root, run, sys, usr, var")
```
This actually should *not* be like this. In `chroot-only` confinement, there should **only** be `/bin` and `/nix` and even `/run` is **not** expected to be there since it should be the *most* restricted environment. Having all that stuff in there means there is certainly something wrong (and already in the first place, since `/run` is in there, which was introduced in via64556974b6674338a227d559975c9b4710f6e751, maybe @andir can explain why).

martinetd replied:
There might be a misunderstanding here -- yes, these directories now listed, but it's definitely not bind mount from the running systems, it's empty directories. There's a new test to make sure the directories are empty and won't bring in new attack surfaces.

Even run is another run directory that has nothing to do with the real /run -- Looking again there are /run/host/os-release and /run/systemd/incoming in there, but it's a small price to pay if that make DynamicUsers and other more commonly used features work...
(Tried adding them to inaccessible paths: systemd/incoming is necessary in setting up the whole private mounts so we service fails to start, but /run/host can be ignored. That appear to be fragments of the systemd "container interface" we probably don't care about: https://systemd.io/CONTAINER_INTERFACE/ - I personally don't see exposing os-release as such a threat, but I agree we might as well hide it. Incoming is just an empty directory once the service is setup so I guess there's not much to hide anyway)

You can spawn a shell with this command to check manually if tests are a bit unwieldy:
```
systemd-run -t -pPrivateMounts=1  -pRuntimeDirectory=systemd-confinement/test -pRootDirectory=/run/systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -pInaccessiblePaths=-+/run/systemd-confinement/test -pInaccessiblePaths=-+/dev -pInaccessiblePaths=-+/sys -pEnvironment=PATH=$(realpath /run/current-system/sw/bin) -- $(realpath /run/current-system/sw/bin/bash)
```

aszlig commented on nixos/tests/systemd-confinement.nix:
```diff
@@ -48,16 +48,22 @@ import ./make-test-python.nix {
       { testScript = ''
           with subtest("full confinement with APIVFS"):
-              machine.fail("chroot-exec ls -l /etc")
+              paths = machine.succeed('chroot-exec ls -Am /etc').strip()
+              assert_eq(paths, "")
```
Again, even with APIVFS, `/etc` should *not* be there if you don't *explicitly* specify it, eg. via `BindPaths` or `BindReadOnlyPaths`.

aszlig reviewed (changes requested):
Thanks a lot for trying to make this work.

However (and maybe I should have commented it better back then) the whole design of the `systemd-confinement` module was so that you get *only* those directories you **explicitly** bind-mount (especially in `chroot-only` mode) so that at the filesystem level the attack surface is as low as possible. This is one of the distinguishing factors between `systemd-confinement` and using all kinds of `Protect*` options.

Just adding those paths to the test surely makes the test succeed, but in this case the implementation is at fault, not the test.

ju1m commented:
@aszlig, but… those directories are **empty** (as commented and checked by the test, and I'll make it even more obvious by using `rmdir`) and [they are forced on us by systemd](https://discourse.nixos.org/t/hardening-systemd-services/17147/18), so much so that in the tests, because each `chroot-exec` starts a socket activated service they are (re)created by each `chroot-exec`.
The only reason why those directories were not present before is because `TemporaryFileSystem=` somehow prevents them from being created by systemd, which only raises ["an ugly error and not a hard failure"](https://github.com/NixOS/nixpkgs/pull/57519#issue-419986659).
So I don't understand your concern here.
It should be possible to `rmdir` those empty directories with an `ExecStartPre=+`, but I don't see what additional security this would bring. To the contrary, doing so would then let a `BindPaths=` to a subpath creates them as an ancestor directory, which uses `UMask=`, which might raise a security concern by setting too relaxed permissions.

aszlig commented:
> @aszlig, but… those directories are **empty** (as commented and checked by the test, and I'll make it even more obvious by using `rmdir`) and [they are forced on us by systemd](https://discourse.nixos.org/t/hardening-systemd-services/17147/18)

Oh, saw that "Check that they're empty" comment and skipped over it because it looked like a todo-comment, sorry X-D

If that's the case then it's not a big issue, though it is indeed very ugly :-/

Will do a second (and this time full) review with testing later when awake.

aszlig commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -98,8 +98,9 @@ in {
-          TemporaryFileSystem = "/";
+          RuntimeDirectory = [ "confinement/${mkPathSafeName name}" ];
+          RootDirectory = lib.mkForce "/run/confinement/${mkPathSafeName name}";
+          InaccessiblePaths = [ "-+/run/confinement/${mkPathSafeName name}" ];
```
Urgh... this sucks :unamused: - we should probably add a subtest to make sure that stuff in `/run` also is as minimal as possible and only contains `/run/systemd/incoming`.

ju1m replied:
Some services (`tor`, `transmission`, `gonic`, `navidrome`, `biboumi`, `pyload`, `tetrd`, `endlessh`, `shiori`, `freeciv`) are already using a `RuntimeDirectory=` as `RootDirectory=`, `mkForce` is here to switch to `systemd-confinement`'s root.

aszlig replied:
> `mkForce` is here to switch to `systemd-confinement`'s root.

The issue I see with this is that it makes it non-obvious. If I set `RootDirectory` for a service and get a different one, I'd be very much confused. Without the `mkForce` I'd get an evaluation error that informs me about conflicting values.

So what's the whole motivation behind this? Is it so that it's easier to use `systemd-confinement` downstream?

ju1m replied:
So far `systemd-confinement` is [never enabled by default in NixOS](https://discourse.nixos.org/t/nixos-policy-regarding-systemd-confinement/18976), and currently uses `TemporaryFileSystem = "/"` which undermines `RootDirectory=`.
@martinetd and I considered using a `mkOverride`, but `mkForce` being much more often used in `nixos/`, I've kept using it, to indeed make it possible to enable `systemd-confinement` on services which already have a transient `RootDirectory=`, as I don't see the point of using `systemd-confinement` if that's to use anything else than a transient `RootDirectory=`.
On second thought, reusing any transient `RootDirectory=` already set by the service would achieve the same, so I propose to switch to `mkDefault`.

martinetd replied:
Hmm, I had a look at tor and I'm not sure it's going to work so well with confinement anyway? They're adding /etc and /nix/store as BindReadOnlyPaths in their own RootDirectory, so whatever we're binding more precisely won't make the rest of the store go away (didn't try to confine the service but we're not resting existing values as far as I can see)

(They're also setting PrivateTmp so they have a writable tmp directory in there, I'd be surprised if tor actually needs it.. It used to be conditioned to its controlSocket, but that is created in the run dir so woulld probably work without?)

So with that in mind I'd say such services are probably a lost battle, but I also agree it doesn't matter if we use our confinement rootdirectory or the one they set, so `mkDefault` sounds good to me.

aszlig commented on nixos/tests/systemd-confinement.nix:
```diff
@@ -88,7 +103,7 @@ import ./make-test-python.nix {
         testScript = ''
           with subtest("check if symlinks are properly bind-mounted"):
-              machine.fail("chroot-exec test -e /etc")
+              machine.succeed("chroot-exec rmdir /etc")
```
:+1: This actually is better than before, since `machine.fail` could fail on unrelated things.

aszlig commented on nixos/tests/systemd-confinement.nix:
```diff
@@ -80,6 +86,15 @@ import ./make-test-python.nix {
               machine.fail("chroot-exec touch /bin/test")
         '';
       }
+      { config.serviceConfig.DynamicUser = true;
```
I'd also do this same test for the `full-apivfs` and `chroot-only` modes, since we really want to make sure that `DynamicUser` behaves the same. One difference when using `chroot-only` is for example that there is a `/var/tmp` directory when using `DynamicUser`.

aszlig reviewed (commented):
Thanks for the clarification... while I still don't like the garbage empty directories, I think this is fine and having support for `DynamicUser` is something I wanted to have for a long time.

There also seems to be an unrelated commit (765a09e53257149a7470ecf9cc5b988cecc6860c) in this branch, was it pushed by accident?

One thing I'd like to see before this is merged are tests for the `ProtectSystem` option and more thorough tests for the `DynamicUser` case (I'd probably just parametrise the existing tests), because using either of these options should not increase the attack surface and *if* that's the case the test should fail.

ju1m commented:
> There also seems to be an unrelated commit ([765a09e](https://github.com/NixOS/nixpkgs/commit/765a09e53257149a7470ecf9cc5b988cecc6860c)) in this branch, was it pushed by accident?

Nope, this is a change following [martinetd's comment](https://github.com/NixOS/nixpkgs/pull/289593#discussion_r1493726108) to overcome the many systemd services which do not use a list for `RuntimeDirectory=` even though they should, to let users or here `systemd-confinement` add their own. Note that it's not removing duplicates though, which is not a problem for `RuntimeDirectory=`-like options, but may be for `ExecStartPre=`-like options. Anyway mixing list and non-list values in a systemd option is currently not supported, so this would not take users by surprise. I'm not sure if it requires a release note or not because it's quite a corner case.

aszlig commented:
> I'm not sure if it requires a release note or not because it's quite a corner case.

IMHO this definitely requires a release note and I would certainly move this to a different pull request because it is not a requirement for supporting `ProtectSystem`/`DynamicUser` and rather a general nice-to-have. Don't get me wrong: I'm very much in support for this change, since I personally had do work around the list/non-list limitation several times, but it's off-topic here in this pull request.

aszlig commented on nixos/tests/systemd-confinement.nix:
```diff
@@ -80,6 +86,77 @@ import ./make-test-python.nix {
+          /tmp
+          /usr
+          /var
+          /var/tmp
```
This should not be there, nor should it be writable.

ju1m replied:
This is enforced by systemd [when `DynamicUser=true`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=):
> If DynamicUser= is enabled, RemoveIPC= and PrivateTmp= are implied (and cannot be turned off).

If need be it can still be disabled with:
```
InaccessiblePaths= [ "-+/tmp" "-+/var/tmp" ];
```
Should I add an option to control that?

aszlig replied:
> Should I add an option to control that?

Hm, difficult to say since then explicitly providing `PrivateTmp` would then result in a no-op. On the other hand when using `systemd-confinement` with `chroot-only` mode, I pretty much expect that there is not a single directory that is writable. For example I usually start with `{ binSh = null; mode = "chroot-only"; }` as the most restrictive environment and then only lift those restrictions when absolutely necessary.

Having a single writable directory there would already implicitly lift those restrictions and the documentation you linked about why `PrivateTmp` are forced-on by default only make sense if you assume that `/tmp` and `/var/tmp` are *already there* which in this case they are not.

Adding stuff to `InaccessiblePaths` also makes it confusing, since we deviate from the systemd documentation. Additionally, "emulating" `PrivateTmp` on our side would introduce unnecessary complexity.

So maybe we should instead document on our side that if you want the *most* restrictive environment (from a filesystem perspective) possible, you have to use dedicated users.

> Should I add an option to control that?

Good question... I usually try to only introduce options whenever really necessary and in this case people can just add `InaccessiblePaths`, but given that this makes it even more verbose to start with the most restrictive configuration, a shortcut could make sense. Additionally, having such a shortcut would also mean that we can explicitly test this scenario.

Better ideas? Also, cc @martinetd on this.

martinetd replied:
Hmm, I'm also of the opinion that chroot-only shouldn't have any writable directory by default.

OTOH, adding tmp paths to InaccessiblePaths without a way to unset it is a bit annoying -- that can be overridden with mkForce on the service's InaccessiblePaths but it's not practical to just remove a single element... Which leaves the hard part of "how".

Thinking out loud:
 - plain new option; I don't really like that as it increases the confinement API which is hard to grasp enough as it is
 - adding new modes e.g. chroot-only/apivfs/tmp; I'm not sure about that either (would it base off apivfs or chroot-only? and it's not really logical)
 - we could look if the user explicitly set ServiceConfig.PrivateTmp and add /tmp+/var/tmp to inaccesiblepaths if and only if PrivateTmp isn't true:
    - That basically overrides the default that DynamicUser implies PrivateTmp, but the rationale in the man page is that it doesn't want the service to leave files behind, which is still respected
    - A user that wants either tmp directory can explicitly request privatetmp and eventually mask the other path if they want to
    - Downside is that it's not as obvious/discoverable as a brand new option
 - Open an issue @ systemd upstream to make PrivateTmp not create the directories if they don't exist
   - That's probably the cleanest way forward: A user can somehow create the directories if they want them
   - OTOH I think it'll be a tough sell to systemd devs, and will take time to be available for us.

I'm honestly not sure what's best here... Kludging over ServiceConfig.PrivateTmp is probably a good compromise even if it's a bit ugly?

aszlig replied:
> * e could look if the user explicitly set ServiceConfig.PrivateTmp and add /tmp+/var/tmp to inaccesiblepaths if and only if PrivateTmp isn't true:

One issue with this is that if you use eg. `systemd.packages`, it could include `PrivateTmp` and we have no way to reliably detect this at eval-time (well, except IFD of course, but...).

ju1m replied:
I explored option 3. (piggybacking on `PrivateTmp=`), tell me if it's worth pursuing/keeping, at least until option 4. (modifying `systemd`) becomes a reality.

ju1m replied:
Note that I had to set `InaccessiblePaths`= through `config.systemd.packages` because `serviceConfig` options cannot depend on each others without causing some obscure infinite recursion.

martinetd replied:
I'm not fond of requiring users to explicitly pull in confinement.dynamicUser; my interest in this PR lies in that existing services become much easier to use as is now so adding more knobs somewhat murks the point. I agree an assertion + separate option makes is stand out more and will lead to less unexpected behaviors, but I expect confinement to give no writable directory more than I expect a /tmp to be available, so I think I would just find this annoying in practice.

As for the default of PrivateTmp "becoming" false I don't think that's much of a problem -- at very least:
- this won't break anything, because this only impacts services with dynamicuser enabled (I haven't found any other setting that implies it), and it's not possible to use it now, so no use case is broken
- when enabling it the user will test the service and I think it'll be obvious that /tmp is missing from any error that pops up (I don't consider confinement to be something that can be enabled and trusted to work without testing); and I find the documentation julm added to be sufficient in understanding what happened (but then again, I was here in this PR, so not in a position to judge how surprising that really is)

So I'm in favor of the current solution, but we can keep discussing if you really prefer having a separate option.

aszlig replied:
Hm, searching on GitHub for [chroot-only language:Nix](https://github.com/search?q=chroot-only+language%3ANix&type=code) turns out that the chroot-only mode seems to be only used by a handful users (at least those making their source code available), so I guess the impact could be rather low, since switching those services to `DynamicUser` is also a conscious decision.

martinetd replied:
(note my /etc/nixos repository is [public](https://cgit.notk.org/asmadeus/nixos-config.git/), but not on github -- and that's also true for julm (not giving a link without his permission)... I'm afraid the subset of folks using chroot-only might also be biased towards not publishing "private-ish" stuff on this platform :/ But yes, the switch back to DynamicUser is a conscious decision)

aszlig commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -183,6 +182,14 @@ in {
+      # If DynamicUser= is enabled, PrivateTmp= is implied (and cannot be turned off).
+      # but with "chroot-only" the tmp directories are not expected,
+      # So disable them unless PrivateTmp=true is explicitely set **in the NixOS configuration**.
+      ${lib.optionalString (cfg.confinement.mode == "chroot-only" && !cfg.serviceConfig.PrivateTmp) ''
```
This also should not depend on `chroot-only`, even though it wouldn't increase the attack surface on `full-apivfs`.

ju1m replied:
I've updated accordingly.

ju1m commented:
Any problem remaining with this PR before asking a committer to merge it? or some concern raised that I've overlooked? Please let me know :)

aszlig commented:
> Any problem remaining with this PR before asking a committer to merge it?

I'll merge this as soon as all points are addressed, so that we don't have regressions and/or surprising behaviour by deviating from the upstream documentations without users being aware of that.

> or some concern raised that I've overlooked? Please let me know :)

We still have the issue that we now deviate from the systemd documentation and so far nobody has either commented on [my proposal](https://github.com/NixOS/nixpkgs/pull/289593#discussion_r1501701403) (the "you" in that comment was explicitly plural) and/or come up with a better idea.

aszlig commented:
@ju1m: Are you okay with me adding commits to your branch?

ju1m commented:
@aszlig, certainly, sorry for the delay, do I have to allow something somewhere?

ju1m commented:
> We still have the issue that we now deviate from the systemd documentation and so far nobody has either commented on [my proposal](https://github.com/NixOS/nixpkgs/pull/289593#discussion_r1501701403) (the "you" in that comment was meant in plural) and/or come up with a better idea.

Ah, sorry I though my comment about piggybacking on `PrivateTmp=` was clear enough. So, to be more precise, because I was able to depend on `serviceConfig.PrivateTmp` to set `serviceConfig.InaccessiblePaths` (via `config.systemd.packages`) I did not feel the need to add a `confinement.dynamicUser`, so I just followed the part about documenting, [in the option](https://github.com/ju1m/nixpkgs/blob/849490ce3ad79f30aca8c71a02b5843acaa3f5ed/nixos/modules/security/systemd-confinement.nix#L89-L94) and in [the `InaccessiblePaths=` config](https://github.com/ju1m/nixpkgs/blob/849490ce3ad79f30aca8c71a02b5843acaa3f5ed/nixos/modules/security/systemd-confinement.nix#L186-L191):
- `systemd-confinement`'s default behavior: recreating `PrivateTmp=false`,
- and how to get systemd's default behavior: by specifying `serviceConfig.PrivateTmp=true`, note the `serviceConfig` to be very specific that it has to be set at NixOS-level and not `config.systemd.packages`' level.

Anecdotically, I was able to understand what to expect and what to do from that new documentation after forgetting about all this for two weeks.
Yet, let me know if your concern about this is not yet solved in a satisfactory manner?

ju1m commented:
- [X] Rebase after conflict in `nixos/doc/manual/release-notes/rl-2405.section.md`

aszlig commented:
I haven't forgotten about this, but I've fallen ill and am on the mend.

aszlig committed:
nixos/tests/confinement: Re-add description attr

The reason why I originally used the "description" attribute was that it
can be easily used to parametrise the tests so that we can specify
common constraints and apply it across a number of different
configurations.

When porting the tests to Python, the description attribute was replaced
by inlining it into the Python code, most probably because it was easier
to do in bulk since using Nix to generate the subtest parts would be
very complicated to do since we also had to please Black (a Python code
formatter that we no longer use in test scripts).

Since we now also want to support DynamicUser in systemd-confinement,
the need to parametrise the tests became apparent again because it's now
easier to refactor our subtests to run both with *and* without
DynamicUser set to true.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig committed:
nixos/tests/confinement: Move to dedicated dir

When experimenting on ways how to refactor the test, I wrote a
significant enough amount of Python to warrant a dedicated Python file.

This commit is mainly to prepare for that and make it easier to track
renames.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig committed:
nixos/tests/confinement: Run test probes in Python

So far the architecture for the tests was that we would use a systemd
socket unit using the Accept option to start a small shell process where
we can pipe commands into by connecting to the socket created by the
socket unit.

This is unnecessary since we can directly use the code snippets from the
individual subtests and systemd will take care of checking the return
code in case we get any assertions[^1].

Another advantage of this is that tests now run in parallel, so we can
do rather expensive things such as looking in /nix to see whether
anything is writable.

The new assert_permissions() function is the main driver behind this and
allows for a more fine-grained way to check whether we got the right
permissions whilst also ignoring irrelevant things such as read-only
empty directories.

Our previous approach also just did a read-only check, which might be
fine in full-apivfs mode where the attack surface already is large, but
in chroot-only mode we really want to make sure nothing is every
writable.

A downside of the new approach is that currently the unit names are
numbered via lib.imap1, which makes it annoying to track its definition.

[^1]: Speaking of assertions, I wrapped the code to be run with pytest's
      assertion rewriting, so that we get more useful AssertionErrors.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig commented on nixos/modules/security/systemd-confinement.nix:
```diff
@@ -98,8 +105,11 @@ in {
       in lib.mkIf config.confinement.enable {
         serviceConfig = {
-          RootDirectory = "/var/empty";
-          TemporaryFileSystem = "/";
```
I've refactored the NixOS test and while working on parametrising the subtests over root/static-user/dynamic-user I found that when using a static user, the root directory is now writable. This is obviously not what we want and setting `TemporaryFileSystem = "/";` here avoids this for both static and dynamic users.

So before I simply re-introduce this option: Is there a particular reason why you removed this?

aszlig replied:
I just pushed the parametrised subtests so you can reproduce the issue.

ju1m replied:
Oh, good catch! [I removed]((https://discourse.nixos.org/t/hardening-systemd-services/17147/14)) `TemporaryFileSystem = [ "/" ]` because [`ProtectSystem=` invalidates `TemporaryFileSystem=/`](https://github.com/systemd/systemd/issues/18999).
But then I did not notice that `/` was no longer necessarily owned by `root` but by the service's user, which is not a problem when `ProtectSystem=strict` (which is implied by `DynamicUser=true`), as `/` is then mounted read-only, but becomes a problem when only using `User=` (as it does not imply `ProtectSystem=strict`).

I can't explore solutions to that problem yet, but here are the details:

Using a `RuntimeDirectory=` as `RootDirectory=` has the side effect of using the service's user as the owner of `/`:
```console
$ sudo systemd-run -pUser=999 -P -pPrivateMounts=1 -pRuntimeDirectory=systemd-confinement/test -pRootDirectory=/run/systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -ld /; mkdir -v /test"
Running as unit: run-u3733.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -ld /
drwxr-xr-x 11 999 999 220 Apr 28 13:23 /
+ mkdir -v /test
mkdir: created directory '/test'
```

Whereas using `TemporaryFileSystem = [ "/" ]` has the side effect of making `root` the owner of `/`:
```console
$ sudo systemd-run -P -pUser=999 -pTemporaryFileSystem=/ -pPrivateMounts=1 -pRuntimeDirectory=systemd-confinement/test -pRootDirectory=/run/systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -ld /; mkdir -v /test"
Running as unit: run-u3740.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -ld /
drwxr-xr-x 7 0 0 140 Apr 28 13:28 /
+ mkdir -v /test
mkdir: cannot create directory '/test': Permission denied
```

The problem is not detected when using `DynamicUser=true` because it implies `ProtectSystem=strict`, making `/` read-only:
```console
$ sudo systemd-run -P -pUser=999 -pProtectSystem=strict -pPrivateMounts=1 -pRuntimeDirectory=systemd-confinement/test -pRootDirectory=/run/systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -ld /; mkdir -v /test"
Running as unit: run-u3742.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -ld /
drwxr-xr-x 11 999 999 220 Apr 28 13:29 /
+ mkdir -v /test
mkdir: cannot create directory '/test': Read-only file system
```

aszlig replied:
> Oh, good catch! I removed `TemporaryFileSystem = [ "/" ]` because [`ProtectSystem=` invalidates `TemporaryFileSystem=/`](https://github.com/systemd/systemd/issues/18999).

So in theory we should just be able to leave `TemporaryFileSystem = "/";` and it should work for both cases since it will not change the behaviour for the `ProtectSystem`/`DynamicUser` case, do I understand this correctly?

ju1m replied:
@aszlig, alas, though `TemporaryFileSystem=/` is (currently) compatible to some extent with `ProtectSystem=strict` (implied by `DynamicUser=true`), it is however incompatible with both `ProtectSystem=full` or `ProtectSystem=true`:

```console
$ sudo systemd-run -P -pTemporaryFileSystem=/ -pProtectSystem=full -pPrivateMounts=1 -pRootDirectory=/run/systemd-confinement/test -pRuntimeDirectory=systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath -e /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -la /; mkdir -v /test"
Running as unit: run-u2785.service

$ systemctl status run-u2785.service
× run-u2785.service - /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test
     Loaded: loaded (/run/systemd/transient/run-u2785.service; transient)
  Transient: yes
     Active: failed (Result: exit-code) since Wed 2024-05-01 21:41:44 CEST; 18s ago
   Duration: 10ms
    Process: 195345 ExecStart=/nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test (code=exited, status=226/NAMESPACE)
   Main PID: 195345 (code=exited, status=226/NAMESPACE)
         IP: 0B in, 0B out
         IO: 0B read, 0B written
        CPU: 6ms

mai 01 21:41:44 oignon systemd[1]: Started /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test.
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed to set up mount namespacing: /run/systemd-confinement/test/etc: No such file or directory
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed at step NAMESPACE spawning /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash: No such file or directory
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Main process exited, code=exited, status=226/NAMESPACE
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Failed with result 'exit-code'.

$ sudo systemd-run -P -pTemporaryFileSystem=/ -pProtectSystem=true -pPrivateMounts=1 -pRootDirectory=/run/systemd-confinement/test -pRuntimeDirectory=systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath -e /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -la /; mkdir -v /test"
Running as unit: run-u2786.service

$ systemctl status run-u2786.service
× run-u2786.service - /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test
     Loaded: loaded (/run/systemd/transient/run-u2786.service; transient)
  Transient: yes
     Active: failed (Result: exit-code) since Wed 2024-05-01 21:41:51 CEST; 48min ago
   Duration: 12ms
    Process: 195363 ExecStart=/nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test (code=exited, status=226/NAMESPACE)
   Main PID: 195363 (code=exited, status=226/NAMESPACE)
         IP: 0B in, 0B out
         IO: 0B read, 0B written
        CPU: 10ms

mai 01 21:41:51 oignon systemd[1]: Started /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test.
mai 01 21:41:51 oignon (bash)[195363]: run-u2786.service: Failed to set up mount namespacing: /run/systemd-confinement/test/usr: No such file or directory
mai 01 21:41:51 oignon (bash)[195363]: run-u2786.service: Failed at step NAMESPACE spawning /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash: No such file or directory
mai 01 21:41:51 oignon systemd[1]: run-u2786.service: Main process exited, code=exited, status=226/NAMESPACE
mai 01 21:41:51 oignon systemd[1]: run-u2786.service: Failed with result 'exit-code'.

mai 01 21:41:44 oignon systemd[1]: Started /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash -xc PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin; ls -la /; mkdir -v /test.
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed to set up mount namespacing: /run/systemd-confinement/test/etc: No such file or directory
mai 01 21:41:44 oignon (bash)[195345]: run-u2785.service: Failed at step NAMESPACE spawning /nix/store/rnzr5gb65b4f5xk7dmljk93km9wpsll6-bash-interactive-5.2-p15/bin/bash: No such file or directory
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Main process exited, code=exited, status=226/NAMESPACE
mai 01 21:41:44 oignon systemd[1]: run-u2785.service: Failed with result 'exit-code'
```

The goal is to deny write access to `/` to non-root users,
- `TemporaryFileSystem=/` gives us that through the ownership of `/` by `root` (instead of the service's user inherited from `RuntimeDirectory=`).
- `ProtectSystem=strict` gives us that by mounting `/` read-only (while keeping its ownership to the service's user).

To avoid the incompatibilities of `TemporaryFileSystem=/` mentionned above, I suggest to mount `/` read-only in all cases with `ReadOnlyPaths = [ "+/" ]`:

```console
$ sudo systemd-run -P -pReadOnlyPaths=+/ -pUser=999 -pPrivateMounts=1 -pRootDirectory=/run/systemd-confinement/test -pRuntimeDirectory=systemd-confinement/test -pUMask=066 -pBindReadOnlyPaths=/nix/store -- $(realpath -e /bin/sh) -xc "PATH=$(dirname $(realpath -e $(which ls))); ls -la /; mkdir -v /test"
Running as unit: run-u2780.service
+ PATH=/nix/store/59mv7mybz16jjvnkz4z5shqgaqm4k009-coreutils-full-9.3/bin
+ ls -la /
total 0
drwxr-xr-x  11 999 999  220 May  1 19:40 .
drwxr-xr-x  11 999 999  220 May  1 19:40 ..
drwxr-xr-x  21   0   0 3800 May  1 16:52 dev
drwxr-xr-x   2   0   0   40 May  1 19:40 etc
drwxr-xr-x   3   0   0   60 May  1 19:40 nix
dr-xr-xr-x 321   0   0    0 May  1 19:40 proc
drwxr-x---   2   0   0   40 May  1 19:40 root
drwxrwxrwt   5   0   0  100 May  1 19:40 run
dr-xr-xr-x  13   0   0    0 Apr 29 18:38 sys
drwxr-xr-x   2   0   0   40 May  1 19:40 usr
drwxr-xr-x   2   0   0   40 May  1 19:40 var
+ mkdir -v /test
mkdir: cannot create directory '/test': Read-only file system
```
I *guess* this would require at least two changes to the current tests:
1. to no longer expect `root` to be able to write to some paths (like `/bin`) (at least not without first remounting `/` in read-write mode).
2. to no longer expect non-`root` users to fail to write to certain paths with a "permission denied" error code, but with a "read-only file system" error code.

aszlig replied:
> it is however incompatible with both `ProtectSystem=full` or `ProtectSystem=true`:

ah, good to know, will add parametrised tests for those options as well.

> To avoid the incompatibilities of `TemporaryFileSystem=/` mentionned above, I suggest to mount `/` read-only in all cases with `ReadOnlyPaths = [ "+/" ]`:

:+1: - will implement that later today or tomorrow.

aszlig committed:
nixos/tests/confinement: Parametrise subtests

This is to make sure that we test all of the DynamicUser/User/Group and
PrivateTmp options in a uniform way. The reason why we need to do this
is because we recently introduced support for the DynamicUser option and
since there are some corner cases where we might end up with more
elevated privileges (eg. writable directories in some cases), we want to
make sure that the environment is as restrictive as with a static
User/Group assignment.

I also removed various checks that try to os.chown(), since with our new
recursive checker those are redundant.

Signed-off-by: aszlig <aszlig@nix.build>

aszlig committed:
nixos/systemd-confinement: Make / read-only

Our more thorough parametrised tests uncovered that with the changes for
supporting DynamicUser, we now have the situation that for static users
the root directory within the confined environment is now writable for
the user in question.

This is obviously not what we want and I'd consider that a regression.
However while discussing this with @ju1m and my suggestion being to
set TemporaryFileSystem to "/" (as we had previously), they had an even
better idea[1]:

> The goal is to deny write access to / to non-root users,
>
>   * TemporaryFileSystem=/ gives us that through the ownership of / by
>     root (instead of the service's user inherited from
>     RuntimeDirectory=).
>   * ProtectSystem=strict gives us that by mounting / read-only (while
>     keeping its ownership to the service's user).
>
> To avoid the incompatibilities of TemporaryFileSystem=/ mentioned
> above, I suggest to mount / read-only in all cases with
> ReadOnlyPaths = [ "+/" ]:
>
>   ...
>
> I guess this would require at least two changes to the current tests:
>
>   1. to no longer expect root to be able to write to some paths (like
>      /bin) (at least not without first remounting / in read-write
>      mode).
>   2. to no longer expect non-root users to fail to write to certain
>      paths with a "permission denied" error code, but with a
>      "read-only file system" error code.

I like the solution with ReadOnlyPaths even more because it further
reduces the attack surface if the user is root. In chroot-only mode this
is especially useful, since if there are no other bind-mounted paths
involved in the unit configuration, the whole file system within the
confined environment is read-only.

[1]: https://github.com/NixOS/nixpkgs/pull/289593#discussion_r1586794215

Signed-off-by: aszlig <aszlig@nix.build>

aszlig reviewed (approved):
@ju1m: Implemented the suggested changes (`ReadOnlyPaths = [ "+/" ]`) and rebased against `master`. Please review and let me know if the implemented changes are fine.

aszlig commented:
@ofborg test systemd-confinement

ju1m commented:
@aszlig thanks!
- [X] `nix -L build -f . nixosTests.systemd-confinement` passed for me
- [X] My confinement of `services.miniflux` using `DynamicUser=true` worked.
- [X] `systemd.services.public-inbox-init.confinement = true` fails at runtime with the usual cryptic error messages due to [`RootDirectory = "/var/empty"; TemporaryFileSystem = "/";`](https://github.com/NixOS/nixpkgs/blob/9fb582619d1397a8cb64cfe28896a064fa1b33cb/nixos/modules/services/mail/public-inbox.nix#L111):
```
Failed to create /var/empty/usr: Operation not permitted
Failed to create destination mount point node '/var/empty/bin/sh': No such file or directory
Failed to mount /nix/store/gdc7yg4hral8i89p9w4gi6496w1p505z-dash-0.5.12/bin/dash to /var/empty/bin/sh: No such file or directory
public-inbox-init.service: Failed to set up mount namespacing: /var/empty/bin/sh: No such file or directory
public-inbox-init.service: Failed at step NAMESPACE spawning /nix/store/ddhngjm9bahdzgimx6bppkzzgd3hjbas-unit-script-public-inbox-init-start/bin/public-inbox-init-star>
```

Maybe we should:
1. Remove `lib.mkDefault` from `systemd-confinement`'s `RootDirectory=` to get a build-time error, or maybe use `lib.mkForce`?
2. Add an assertion in `systemd-confinement` checking that there is no `"/"` in `TemporaryFileSystem`'s list.
3. Make `services.public-inbox` only set `RootDirectory=` and `TemporaryFileSystem=/` when `confinement` is not enabled. I'll prepare a PR for that, though I'll much prefer to just have `confinement` enabled by default on that service…

aszlig commented:
> Maybe we should:
> 
> 1. Remove `lib.mkDefault` from `systemd-confinement`'s `RootDirectory=` to get a build-time error, or maybe use `lib.mkForce`?

Is the issue the `RootDirectory` option or is it because we now set `ReadOnlyPaths`?

> 2. Add an assertion in `systemd-confinement` checking that there is no `"/"` in `TemporaryFileSystem`'s list.
> 3. Make `services.public-inbox` only set `RootDirectory=` and `TemporaryFileSystem=/` when `confinement` is not enabled. I'll prepare a PR for that, though I'll much prefer to just have `confinement` enabled by default on that service…

We also have a fourth option that is similar to the second: Check whether there is `/` in `TemporaryFileSystem` and if that's the case[^1] we remove it from `ReadOnlyPaths`.

This however is one particular corner case which we're optimising for in the confinement module, which I'm not sure would be the right place. If it's something that would generally make sense in conjunction with the confinement module, I'd say it's fine but I have a hard time coming up with such a scenario.

So I'd probably opt for the third option and only set those options without confinement enabled.

Is there a feature (or bugfix) that's missing in the confinement module which is preventing it from being enabled by default on that module?

[^1]: This is me *assuming* the failure is due to the `ReadOnlyPaths` setting, I haven't verified this.

ju1m commented:
> > 1. Remove `lib.mkDefault` from `systemd-confinement`'s `RootDirectory=` to get a build-time error, or maybe use `lib.mkForce`?
> 
> Is the issue the `RootDirectory` option or is it because we now set `ReadOnlyPaths`?

`public-inbox`'s `RootDirectory=/var/empty` is the problem.
It's unrelated to `ReadOnlyPaths=`.
It's a (silent) collision between `public-inbox` and `systemd-confinement` both  setting their own `RootDirectory=`, hence my suggestion to at least remove `lib.mkDefault` to make the collision visible at build-time.
I've pushed 3. in https://github.com/NixOS/nixpkgs/pull/311181 which tests `public-inbox`'s confinement and fixes it with the changes in this PR.

> > 2. Add an assertion in `systemd-confinement` checking that there is no `"/"` in `TemporaryFileSystem`'s list.
> > 3. Make `services.public-inbox` only set `RootDirectory=` and `TemporaryFileSystem=/` when `confinement` is not enabled. I'll prepare a PR for that, though I'll much prefer to just have `confinement` enabled by default on that service…
> 
> We also have a fourth option that is similar to the second: Check whether there is `/` in `TemporaryFileSystem` and if that's the case[1](#user-content-fn-1-a52fa1de105644658b8dac320abd0535) we remove it from `ReadOnlyPaths`.

Alas, changing `serviceConfig` settings depending on the **content** of other `serviceConfig` settings is not possible (infinite recursion).
That's why we had to resort to the `config.systemd.packages` hack for setting `InaccessiblePaths=` when `serviceConfig.PrivateTmp` is `false`.

> This however is a corner case which we'd be optimising for in the confinement module, which I'm not sure would be the right place. If it's something that would generally make sense in conjunction with the confinement module I'd say it's fine, but I have a hard time coming up with such a scenario.
I agree, I was just suggesting two more build-time checks, not to automatically change the configuration.
> 
> So I'd probably opt for the third option and only set those options without confinement enabled.

Yes, I've done 3. I leave it to you whether to do 1. and 2. or not, as this PR is already very good to merge for me, and having it merged before 2024-05-22 would bring it into `nixos-24.05`.

> Is there a feature (or bugfix) that's missing in the confinement module which is preventing it from being enabled by default on that module?

Yes, in the PR I had to change `RootDirectory=/var/empty` for `RootDirectory = "/run/confinement/public-inbox-${srv}";` and remove `TemporarySystem=/` which I set before because it was what `systemd-confinement` was using anyway.

aszlig commented:
> Yes, I've done 3. I leave it to you whether to do 1. and 2. or not, as this PR is already very good to merge for me, and having it merged before 2024-05-22 would bring it into `nixos-24.05`.

:+1: I'll implement point 1, rebase and merge. Thanks for your input and the work done here :-)

aszlig committed:
nixos/confinement: Use prio 100 for RootDirectory

One of the module that already supports the systemd-confinement module
is public-inbox. However with the changes to support DynamicUser and
ProtectSystem, the module will now fail at runtime if confinement is
enabled (it's optional and you'll need to override it via another
module).

The reason is that the RootDirectory is set to /var/empty in the
public-inbox module, which doesn't work well with the InaccessiblePaths
directive we now use to support DynamicUser/ProtectSystem.

To make this issue more visible, I decided to just change the priority
of the RootDirectory option definiton the default override priority so
that whenever another different option is defined, we'll get a conflict
at evaluation time.

Signed-off-by: aszlig <aszlig@nix.build>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-policy-regarding-systemd-confinement/18976/2
